src/playlist-loader.js

```diff
@@ -5,376 +5,384 @@
+  * between the media sequence numbers in the original and updated
+  * playlists.
+  * @return a list of merged segment objects
+  */
+const updateSegments = function(original, update, offset) {
+  let result = update.slice();
+  let length;
+  let i;
+
+  offset = offset || 0;
+  length = Math.min(original.length, update.length + offset);
+
+  for (i = offset; i < length; i++) {
+    result[i - offset] = mergeOptions(original[i], result[i - offset]);
+  }
+  return result;
+};
+
+/**
+  * Returns a new master playlist that is the result of merging an
+  * updated media playlist into the original version. If the
+  * updated media playlist does not match any of the playlist
+  * entries in the original master playlist, null is returned.
+  * @param master {object} a parsed master M3U8 object
+  * @param media {object} a parsed media M3U8 object
+  * @return {object} a new object that represents the original
+  * master playlist with the updated media playlist merged in, or
+  * null if the merge produced no change.
+  */
+const updateMaster = function(master, media) {
+  let changed = false;
+  let result = mergeOptions(master, {});
+  let i = master.playlists.length;
+  let playlist;
+
+  while (i--) {
+    playlist = result.playlists[i];
+    if (playlist.uri === media.uri) {
+      // consider the playlist unchanged if the number of segments
+      // are equal and the media sequence number is unchanged
+      if (playlist.segments &&
+          media.segments &&
+          playlist.segments.length === media.segments.length &&
+          playlist.mediaSequence === media.mediaSequence) {
+        continue;
       }
-      return result;
-    },
-
-    PlaylistLoader = function(srcUrl, withCredentials) {
-      var
-        loader = this,
-        dispose,
-        mediaUpdateTimeout,
-        request,
-        playlistRequestError,
-        haveMetadata;
-
-      PlaylistLoader.prototype.init.call(this);
-
-      // a flag that disables "expired time"-tracking this setting has
-      // no effect when not playing a live stream
-      this.trackExpiredTime_ = false;
-
-      if (!srcUrl) {
-        throw new Error('A non-empty playlist URL is required');
+
+      result.playlists[i] = mergeOptions(playlist, media);
+      result.playlists[media.uri] = result.playlists[i];
+
+      // if the update could overlap existing segment information,
+      // merge the two lists
+      if (playlist.segments) {
+        result.playlists[i].segments = updateSegments(playlist.segments,
+                                                      media.segments,
+                                                      media.mediaSequence -
+                                                      playlist.mediaSequence);
       }
+      changed = true;
+    }
+  }
+  return changed ? result : null;
+};
 
-      playlistRequestError = function(xhr, url, startingState) {
-        loader.setBandwidth(request || xhr);
+export default class PlaylistLoader extends Stream {
```

Comment:
Looks like there are a couple refactorings going on simultaneously in this class. Can this be split up into something like:

1. es6 stuff
2. (optional) reorder stuff